File: Sources/SwiftCogCore/Kernels/ExecutiveKernel.swift
import Foundation
import Distributed
import DistributedCluster

public distributed actor ExecutiveKernel: Kernel {
    let kernelId: KernelID
    unowned let system: KernelSystem
    private let customHandler: ((KernelMessage, ExecutiveKernel) async throws -> Void)?

    public init(actorSystem: DefaultDistributedActorSystem, system: KernelSystem, customHandler: ((KernelMessage, ExecutiveKernel) async throws -> Void)? = nil) {
        self.actorSystem = actorSystem
        self.kernelId = KernelID()
        self.system = system
        self.customHandler = customHandler
    }

    public distributed func getKernelId() -> KernelID {
        return self.kernelId
    }

    public distributed func receive(message: KernelMessage) {
        Task {
            if let customHandler = customHandler {
                // Use the custom handler if provided
                try await customHandler(message, self)
            } else {
                // Default behavior: pass the message on
                try await system.emit(message: message, from: self)
            }
        }
    }
} 

File: Sources/server/executive_kernel.py
"""
Executive kernel implementation for the SwiftCog Python server.
"""
from typing import Callable, Optional
import ray
from swiftcog_types import KernelID, KernelMessage


@ray.remote
class ExecutiveKernel:
    """Executive kernel implementation matching the Swift version."""
    
    def __init__(self, custom_handler: Optional[Callable] = None):
        self.custom_handler = custom_handler
        self.kernel_id = KernelID.EXECUTIVE
    
    def get_kernel_id(self) -> KernelID:
        return self.kernel_id
    
    async def receive(self, message: KernelMessage) -> None:
        """Receive and process a message."""
        if self.custom_handler:
            await self.custom_handler(message, self)
        else:
            await self.default_handler(message)
    
    async def default_handler(self, message: KernelMessage) -> None:
        """Default handler that makes executive decisions."""
        response = KernelMessage(
            source_kernel_id=KernelID.EXECUTIVE,
            payload=f"Executive decision made for: '{message.payload}'"
        )
        # Note: Default handlers don't emit - custom handlers handle system communication 